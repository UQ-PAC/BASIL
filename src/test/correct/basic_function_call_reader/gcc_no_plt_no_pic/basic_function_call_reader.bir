00000619: program
000005e3: sub __cxa_finalize(__cxa_finalize_result)
0000061a: __cxa_finalize_result :: out u32 = low:32[R0]

00000389:
00000481: R16 := 0x20000
00000488: R17 := mem[R16 + 8, el]:u64
0000048e: R16 := R16 + 8
00000493: call R17 with noreturn

000005e4: sub __do_global_dtors_aux(__do_global_dtors_aux_result)
0000061b: __do_global_dtors_aux_result :: out u32 = low:32[R0]

0000028b:
0000028f: #3 := R31 - 0x20
00000295: mem := mem with [#3, el]:u64 <- R29
0000029b: mem := mem with [#3 + 8, el]:u64 <- R30
0000029f: R31 := #3
000002a5: R29 := R31
000002ad: mem := mem with [R31 + 0x10, el]:u64 <- R19
000002b2: R19 := 0x20000
000002b9: R0 := pad:64[mem[R19 + 0x30]]
000002bf: when 0:0[R0] goto %000002bd
0000060d: goto %00000352

00000352:
00000355: R0 := 0x1F000
0000035c: R0 := mem[R0 + 0xFC8, el]:u64
00000362: when R0 = 0 goto %00000360
0000060e: goto %00000379

00000379:
0000037c: R0 := 0x20000
00000383: R0 := mem[R0 + 0x28, el]:u64
00000388: R30 := 0x6F0
0000038b: call @__cxa_finalize with return %00000360

00000360:
00000368: R30 := 0x6F4
0000036a: call @deregister_tm_clones with return %0000036c

0000036c:
0000036f: R0 := 1
00000377: mem := mem with [R19 + 0x30] <- 7:0[R0]
0000060f: goto %000002bd

000002bd:
000002c7: R19 := mem[R31 + 0x10, el]:u64
000002ce: R29 := mem[R31, el]:u64
000002d3: R30 := mem[R31 + 8, el]:u64
000002d7: R31 := R31 + 0x20
000002dc: call R30 with noreturn

000005e8: sub __libc_start_main(__libc_start_main_main, __libc_start_main_arg2, __libc_start_main_arg3, __libc_start_main_auxv, __libc_start_main_result)
0000061c: __libc_start_main_main :: in u64 = R0
0000061d: __libc_start_main_arg2 :: in u32 = low:32[R1]
0000061e: __libc_start_main_arg3 :: in out u64 = R2
0000061f: __libc_start_main_auxv :: in out u64 = R3
00000620: __libc_start_main_result :: out u32 = low:32[R0]

000001e4:
0000046b: R16 := 0x20000
00000472: R17 := mem[R16, el]:u64
00000478: R16 := R16
0000047d: call R17 with noreturn

000005e9: sub _fini(_fini_result)
00000621: _fini_result :: out u32 = low:32[R0]

00000020:
00000026: #0 := R31 - 0x10
0000002c: mem := mem with [#0, el]:u64 <- R29
00000032: mem := mem with [#0 + 8, el]:u64 <- R30
00000036: R31 := #0
0000003c: R29 := R31
00000043: R29 := mem[R31, el]:u64
00000048: R30 := mem[R31 + 8, el]:u64
0000004c: R31 := R31 + 0x10
00000051: call R30 with noreturn

000005ea: sub _init(_init_result)
00000622: _init_result :: out u32 = low:32[R0]

00000536:
0000053c: #6 := R31 - 0x10
00000542: mem := mem with [#6, el]:u64 <- R29
00000548: mem := mem with [#6 + 8, el]:u64 <- R30
0000054c: R31 := #6
00000552: R29 := R31
00000557: R30 := 0x590
00000559: call @call_weak_fn with return %0000055b

0000055b:
00000560: R29 := mem[R31, el]:u64
00000565: R30 := mem[R31 + 8, el]:u64
00000569: R31 := R31 + 0x10
0000056e: call R30 with noreturn

000005eb: sub _start(_start_result)
00000623: _start_result :: out u32 = low:32[R0]

000001a5:
000001aa: R29 := 0
000001af: R30 := 0
000001b5: R5 := R0
000001bc: R1 := mem[R31, el]:u64
000001c2: R2 := R31 + 8
000001c8: R6 := R31
000001cd: R0 := 0x1F000
000001d4: R0 := mem[R0 + 0xFD8, el]:u64
000001d9: R3 := 0
000001de: R4 := 0
000001e3: R30 := 0x630
000001e6: call @__libc_start_main with return %000001e8

000001e8:
000001eb: R30 := 0x634
000001ee: call @abort with return %00000610

00000610:
00000611: call @call_weak_fn with noreturn

000005ee: sub abort()


000001ec:
000004ad: R16 := 0x20000
000004b4: R17 := mem[R16 + 0x18, el]:u64
000004ba: R16 := R16 + 0x18
000004bf: call R17 with noreturn

000005ef: sub call_weak_fn(call_weak_fn_result)
00000624: call_weak_fn_result :: out u32 = low:32[R0]

000001f0:
000001f3: R0 := 0x1F000
000001fa: R0 := mem[R0 + 0xFD0, el]:u64
00000200: when R0 = 0 goto %000001fe
00000612: goto %000003c9

000001fe:
00000206: call R30 with noreturn

000003c9:
000003cc: goto @__gmon_start__

000003ca:
00000497: R16 := 0x20000
0000049e: R17 := mem[R16 + 0x10, el]:u64
000004a4: R16 := R16 + 0x10
000004a9: call R17 with noreturn

000005f1: sub deregister_tm_clones(deregister_tm_clones_result)
00000625: deregister_tm_clones_result :: out u32 = low:32[R0]

0000020c:
0000020f: R0 := 0x20000
00000215: R0 := R0 + 0x30
0000021a: R1 := 0x20000
00000220: R1 := R1 + 0x30
00000226: #1 := ~R0
0000022b: #2 := R1 + ~R0
00000231: VF := extend:65[#2 + 1] <> extend:65[R1] + extend:65[#1] + 1
00000237: CF := pad:65[#2 + 1] <> pad:65[R1] + pad:65[#1] + 1
0000023b: ZF := #2 + 1 = 0
0000023f: NF := 63:63[#2 + 1]
00000245: when ZF goto %00000243
00000613: goto %000003ab

000003ab:
000003ae: R1 := 0x1F000
000003b5: R1 := mem[R1 + 0xFC0, el]:u64
000003ba: when R1 = 0 goto %00000243
00000614: goto %000003be

00000243:
0000024b: call R30 with noreturn

000003be:
000003c2: R16 := R1
000003c7: call R16 with noreturn

000005f4: sub frame_dummy(frame_dummy_result)
00000626: frame_dummy_result :: out u32 = low:32[R0]

000002e2:
000002e4: call @register_tm_clones with noreturn

000005f5: sub main(main_argc, main_argv, main_result)
00000627: main_argc :: in u32 = low:32[R0]
00000628: main_argv :: in out u64 = R1
00000629: main_result :: out u32 = low:32[R0]

000002e6:
000002ea: R31 := R31 - 0x10
000002ef: R0 := 0x20000
000002f5: R0 := R0 + 0x38
000002fc: R0 := pad:64[mem[R0, el]:u32]
00000304: mem := mem with [R31 + 0xC, el]:u32 <- 31:0[R0]
00000309: R0 := 0x20000
0000030f: R0 := R0 + 0x34
00000316: R0 := pad:64[mem[R0, el]:u32]
0000031c: #4 := 31:0[R0] - 1
00000321: VF := extend:33[#4 + 1] <> extend:33[31:0[R0]] + 0
00000326: CF := pad:33[#4 + 1] <> pad:33[31:0[R0]] - 0x100000000
0000032a: ZF := #4 + 1 = 0
0000032e: NF := 31:31[#4 + 1]
00000334: when ZF <> 1 goto %00000332
00000615: goto %00000349

00000332:
0000033c: R0 := pad:64[mem[R31 + 0xC, el]:u32]
00000616: goto %0000033e

00000349:
0000034c: R0 := 0
00000350: goto %0000033e

0000033e:
00000342: R31 := R31 + 0x10
00000347: call R30 with noreturn

000005f8: sub register_tm_clones(register_tm_clones_result)
0000062a: register_tm_clones_result :: out u32 = low:32[R0]

0000024d:
00000250: R0 := 0x20000
00000256: R0 := R0 + 0x30
0000025b: R1 := 0x20000
00000261: R1 := R1 + 0x30
00000268: R1 := R1 + ~R0 + 1
0000026e: R2 := 0.63:63[R1]
00000275: R1 := R2 + (R1 ~>> 3)
0000027b: R1 := extend:64[63:1[R1]]
00000281: when R1 = 0 goto %0000027f
00000617: goto %0000038d

0000038d:
00000390: R2 := 0x1F000
00000397: R2 := mem[R2 + 0xFE0, el]:u64
0000039c: when R2 = 0 goto %0000027f
00000618: goto %000003a0

0000027f:
00000287: call R30 with noreturn

000003a0:
000003a4: R16 := R2
000003a9: call R16 with noreturn
