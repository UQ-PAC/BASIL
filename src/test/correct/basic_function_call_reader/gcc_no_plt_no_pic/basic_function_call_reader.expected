var Gamma_R0: bool;
var Gamma_R31: bool;
var Gamma_mem: [bv64]bool;
var Gamma_stack: [bv64]bool;
var R0: bv64;
var R31: bv64;
var mem: [bv64]bv8;
var stack: [bv64]bv8;
const $x_addr: bv64;
axiom ($x_addr == 69652bv64);
const $y_addr: bv64;
axiom ($y_addr == 69656bv64);
function L(memory: [bv64]bv8, index: bv64) returns (bool) {
  (if (index == $y_addr) then (memory_load32_le(memory, $x_addr) == 1bv32) else (if (index == $x_addr) then true else false))
}

function {:bvbuiltin "bvadd"} bvadd32(bv32, bv32) returns (bv32);
function {:bvbuiltin "bvadd"} bvadd33(bv33, bv33) returns (bv33);
function {:bvbuiltin "bvadd"} bvadd64(bv64, bv64) returns (bv64);
function {:bvbuiltin "bvcomp"} bvcomp1(bv1, bv1) returns (bv1);
function {:bvbuiltin "bvcomp"} bvcomp32(bv32, bv32) returns (bv1);
function {:bvbuiltin "bvcomp"} bvcomp33(bv33, bv33) returns (bv1);
function {:bvbuiltin "bvnot"} bvnot1(bv1) returns (bv1);
function gamma_load32(gammaMap: [bv64]bool, index: bv64) returns (bool) {
  (gammaMap[bvadd64(index, 3bv64)] && (gammaMap[bvadd64(index, 2bv64)] && (gammaMap[bvadd64(index, 1bv64)] && gammaMap[index])))
}

function gamma_store32(gammaMap: [bv64]bool, index: bv64, value: bool) returns ([bv64]bool) {
  gammaMap[index := value][bvadd64(index, 1bv64) := value][bvadd64(index, 2bv64) := value][bvadd64(index, 3bv64) := value]
}

function memory_load32_le(memory: [bv64]bv8, index: bv64) returns (bv32) {
  (memory[bvadd64(index, 3bv64)] ++ (memory[bvadd64(index, 2bv64)] ++ (memory[bvadd64(index, 1bv64)] ++ memory[index])))
}

function memory_load8_le(memory: [bv64]bv8, index: bv64) returns (bv8) {
  memory[index]
}

function memory_store32_le(memory: [bv64]bv8, index: bv64, value: bv32) returns ([bv64]bv8) {
  memory[index := value[8:0]][bvadd64(index, 1bv64) := value[16:8]][bvadd64(index, 2bv64) := value[24:16]][bvadd64(index, 3bv64) := value[32:24]]
}

function {:bvbuiltin "sign_extend 1"} sign_extend1_32(bv32) returns (bv33);
function {:bvbuiltin "zero_extend 1"} zero_extend1_32(bv32) returns (bv33);
function {:bvbuiltin "zero_extend 32"} zero_extend32_32(bv32) returns (bv64);
procedure rely();
  modifies mem, Gamma_mem;
  ensures (forall i: bv64 :: (((mem[i] == old(mem[i])) ==> (Gamma_mem[i] == old(Gamma_mem[i])))));
  ensures ((old(memory_load32_le(mem, $x_addr)) == 0bv32) ==> (memory_load32_le(mem, $x_addr) == 0bv32));
  ensures (old(gamma_load32(Gamma_mem, $y_addr)) ==> ((memory_load32_le(mem, $x_addr) == 0bv32) || gamma_load32(Gamma_mem, $y_addr)));

procedure rely_transitive()
  modifies mem, Gamma_mem;
  ensures ((old(memory_load32_le(mem, $x_addr)) == 0bv32) ==> (memory_load32_le(mem, $x_addr) == 0bv32));
  ensures (old(gamma_load32(Gamma_mem, $y_addr)) ==> ((memory_load32_le(mem, $x_addr) == 0bv32) || gamma_load32(Gamma_mem, $y_addr)));
{
  call rely();
  call rely();
}

procedure rely_reflexive()
{
  assert ((memory_load32_le(mem, $x_addr) == 0bv32) ==> (memory_load32_le(mem, $x_addr) == 0bv32));
  assert (gamma_load32(Gamma_mem, $y_addr) ==> ((memory_load32_le(mem, $x_addr) == 0bv32) || gamma_load32(Gamma_mem, $y_addr)));
}

procedure guarantee_reflexive()
  modifies mem, Gamma_mem;
{
  assert (memory_load32_le(mem, $x_addr) == memory_load32_le(mem, $x_addr));
  assert (memory_load32_le(mem, $y_addr) == memory_load32_le(mem, $y_addr));
}

procedure main()
  modifies Gamma_R0, Gamma_R31, Gamma_stack, R0, R31, stack;
  free requires (memory_load8_le(mem, 1892bv64) == 1bv8);
  free requires (memory_load8_le(mem, 1893bv64) == 0bv8);
  free requires (memory_load8_le(mem, 1894bv64) == 2bv8);
  free requires (memory_load8_le(mem, 1895bv64) == 0bv8);
  free requires (memory_load8_le(mem, 69528bv64) == 0bv8);
  free requires (memory_load8_le(mem, 69529bv64) == 0bv8);
  free requires (memory_load8_le(mem, 69530bv64) == 0bv8);
  free requires (memory_load8_le(mem, 69531bv64) == 0bv8);
  free requires (memory_load8_le(mem, 69532bv64) == 0bv8);
  free requires (memory_load8_le(mem, 69533bv64) == 0bv8);
  free requires (memory_load8_le(mem, 69534bv64) == 0bv8);
  free requires (memory_load8_le(mem, 69535bv64) == 0bv8);
  free requires (memory_load8_le(mem, 69536bv64) == 0bv8);
  free requires (memory_load8_le(mem, 69537bv64) == 0bv8);
  free requires (memory_load8_le(mem, 69538bv64) == 0bv8);
  free requires (memory_load8_le(mem, 69539bv64) == 0bv8);
  free requires (memory_load8_le(mem, 69540bv64) == 0bv8);
  free requires (memory_load8_le(mem, 69541bv64) == 0bv8);
  free requires (memory_load8_le(mem, 69542bv64) == 0bv8);
  free requires (memory_load8_le(mem, 69543bv64) == 0bv8);
  free requires (memory_load8_le(mem, 69544bv64) == 0bv8);
  free requires (memory_load8_le(mem, 69545bv64) == 0bv8);
  free requires (memory_load8_le(mem, 69546bv64) == 0bv8);
  free requires (memory_load8_le(mem, 69547bv64) == 0bv8);
  free requires (memory_load8_le(mem, 69548bv64) == 0bv8);
  free requires (memory_load8_le(mem, 69549bv64) == 0bv8);
  free requires (memory_load8_le(mem, 69550bv64) == 0bv8);
  free requires (memory_load8_le(mem, 69551bv64) == 0bv8);
  free requires (memory_load8_le(mem, 69552bv64) == 160bv8);
  free requires (memory_load8_le(mem, 69553bv64) == 5bv8);
  free requires (memory_load8_le(mem, 69554bv64) == 0bv8);
  free requires (memory_load8_le(mem, 69555bv64) == 0bv8);
  free requires (memory_load8_le(mem, 69556bv64) == 0bv8);
  free requires (memory_load8_le(mem, 69557bv64) == 0bv8);
  free requires (memory_load8_le(mem, 69558bv64) == 0bv8);
  free requires (memory_load8_le(mem, 69559bv64) == 0bv8);
  free requires (memory_load8_le(mem, 69560bv64) == 160bv8);
  free requires (memory_load8_le(mem, 69561bv64) == 5bv8);
  free requires (memory_load8_le(mem, 69562bv64) == 0bv8);
  free requires (memory_load8_le(mem, 69563bv64) == 0bv8);
  free requires (memory_load8_le(mem, 69564bv64) == 0bv8);
  free requires (memory_load8_le(mem, 69565bv64) == 0bv8);
  free requires (memory_load8_le(mem, 69566bv64) == 0bv8);
  free requires (memory_load8_le(mem, 69567bv64) == 0bv8);
  free requires (memory_load8_le(mem, 69568bv64) == 160bv8);
  free requires (memory_load8_le(mem, 69569bv64) == 5bv8);
  free requires (memory_load8_le(mem, 69570bv64) == 0bv8);
  free requires (memory_load8_le(mem, 69571bv64) == 0bv8);
  free requires (memory_load8_le(mem, 69572bv64) == 0bv8);
  free requires (memory_load8_le(mem, 69573bv64) == 0bv8);
  free requires (memory_load8_le(mem, 69574bv64) == 0bv8);
  free requires (memory_load8_le(mem, 69575bv64) == 0bv8);
  free requires (memory_load8_le(mem, 69576bv64) == 160bv8);
  free requires (memory_load8_le(mem, 69577bv64) == 5bv8);
  free requires (memory_load8_le(mem, 69578bv64) == 0bv8);
  free requires (memory_load8_le(mem, 69579bv64) == 0bv8);
  free requires (memory_load8_le(mem, 69580bv64) == 0bv8);
  free requires (memory_load8_le(mem, 69581bv64) == 0bv8);
  free requires (memory_load8_le(mem, 69582bv64) == 0bv8);
  free requires (memory_load8_le(mem, 69583bv64) == 0bv8);
  free requires (memory_load8_le(mem, 69584bv64) == 168bv8);
  free requires (memory_load8_le(mem, 69585bv64) == 13bv8);
  free requires (memory_load8_le(mem, 69586bv64) == 1bv8);
  free requires (memory_load8_le(mem, 69587bv64) == 0bv8);
  free requires (memory_load8_le(mem, 69588bv64) == 0bv8);
  free requires (memory_load8_le(mem, 69589bv64) == 0bv8);
  free requires (memory_load8_le(mem, 69590bv64) == 0bv8);
  free requires (memory_load8_le(mem, 69591bv64) == 0bv8);
  free requires (memory_load8_le(mem, 69592bv64) == 0bv8);
  free requires (memory_load8_le(mem, 69593bv64) == 0bv8);
  free requires (memory_load8_le(mem, 69594bv64) == 0bv8);
  free requires (memory_load8_le(mem, 69595bv64) == 0bv8);
  free requires (memory_load8_le(mem, 69596bv64) == 0bv8);
  free requires (memory_load8_le(mem, 69597bv64) == 0bv8);
  free requires (memory_load8_le(mem, 69598bv64) == 0bv8);
  free requires (memory_load8_le(mem, 69599bv64) == 0bv8);
  free requires (memory_load8_le(mem, 69600bv64) == 0bv8);
  free requires (memory_load8_le(mem, 69601bv64) == 0bv8);
  free requires (memory_load8_le(mem, 69602bv64) == 0bv8);
  free requires (memory_load8_le(mem, 69603bv64) == 0bv8);
  free requires (memory_load8_le(mem, 69604bv64) == 0bv8);
  free requires (memory_load8_le(mem, 69605bv64) == 0bv8);
  free requires (memory_load8_le(mem, 69606bv64) == 0bv8);
  free requires (memory_load8_le(mem, 69607bv64) == 0bv8);
  free requires (memory_load8_le(mem, 69608bv64) == 0bv8);
  free requires (memory_load8_le(mem, 69609bv64) == 0bv8);
  free requires (memory_load8_le(mem, 69610bv64) == 0bv8);
  free requires (memory_load8_le(mem, 69611bv64) == 0bv8);
  free requires (memory_load8_le(mem, 69612bv64) == 0bv8);
  free requires (memory_load8_le(mem, 69613bv64) == 0bv8);
  free requires (memory_load8_le(mem, 69614bv64) == 0bv8);
  free requires (memory_load8_le(mem, 69615bv64) == 0bv8);
  free requires (memory_load8_le(mem, 69616bv64) == 20bv8);
  free requires (memory_load8_le(mem, 69617bv64) == 7bv8);
  free requires (memory_load8_le(mem, 69618bv64) == 0bv8);
  free requires (memory_load8_le(mem, 69619bv64) == 0bv8);
  free requires (memory_load8_le(mem, 69620bv64) == 0bv8);
  free requires (memory_load8_le(mem, 69621bv64) == 0bv8);
  free requires (memory_load8_le(mem, 69622bv64) == 0bv8);
  free requires (memory_load8_le(mem, 69623bv64) == 0bv8);
  free requires (memory_load8_le(mem, 69624bv64) == 0bv8);
  free requires (memory_load8_le(mem, 69625bv64) == 0bv8);
  free requires (memory_load8_le(mem, 69626bv64) == 0bv8);
  free requires (memory_load8_le(mem, 69627bv64) == 0bv8);
  free requires (memory_load8_le(mem, 69628bv64) == 0bv8);
  free requires (memory_load8_le(mem, 69629bv64) == 0bv8);
  free requires (memory_load8_le(mem, 69630bv64) == 0bv8);
  free requires (memory_load8_le(mem, 69631bv64) == 0bv8);
  free requires (memory_load8_le(mem, 69632bv64) == 0bv8);
  free requires (memory_load8_le(mem, 69633bv64) == 0bv8);
  free requires (memory_load8_le(mem, 69634bv64) == 0bv8);
  free requires (memory_load8_le(mem, 69635bv64) == 0bv8);
  free requires (memory_load8_le(mem, 69636bv64) == 0bv8);
  free requires (memory_load8_le(mem, 69637bv64) == 0bv8);
  free requires (memory_load8_le(mem, 69638bv64) == 0bv8);
  free requires (memory_load8_le(mem, 69639bv64) == 0bv8);
  free requires (memory_load8_le(mem, 69640bv64) == 8bv8);
  free requires (memory_load8_le(mem, 69641bv64) == 16bv8);
  free requires (memory_load8_le(mem, 69642bv64) == 1bv8);
  free requires (memory_load8_le(mem, 69643bv64) == 0bv8);
  free requires (memory_load8_le(mem, 69644bv64) == 0bv8);
  free requires (memory_load8_le(mem, 69645bv64) == 0bv8);
  free requires (memory_load8_le(mem, 69646bv64) == 0bv8);
  free requires (memory_load8_le(mem, 69647bv64) == 0bv8);
  free ensures (Gamma_R31 == old(Gamma_R31));
  free ensures (R31 == old(R31));
{
  var #4: bv32;
  var CF: bv1;
  var Gamma_#4: bool;
  var Gamma_CF: bool;
  var Gamma_NF: bool;
  var Gamma_VF: bool;
  var Gamma_ZF: bool;
  var NF: bv1;
  var VF: bv1;
  var ZF: bv1;
  lmain:
    R31, Gamma_R31 := bvadd64(R31, 18446744073709551600bv64), Gamma_R31;
    R0, Gamma_R0 := 69632bv64, true;
    R0, Gamma_R0 := bvadd64(R0, 24bv64), Gamma_R0;
    R0, Gamma_R0 := zero_extend32_32(memory_load32_le(mem, R0)), (gamma_load32(Gamma_mem, R0) || L(mem, R0));
    stack, Gamma_stack := memory_store32_le(stack, bvadd64(R31, 12bv64), R0[32:0]), gamma_store32(Gamma_stack, bvadd64(R31, 12bv64), Gamma_R0);
    R0, Gamma_R0 := 69632bv64, true;
    R0, Gamma_R0 := bvadd64(R0, 20bv64), Gamma_R0;
    R0, Gamma_R0 := zero_extend32_32(memory_load32_le(mem, R0)), (gamma_load32(Gamma_mem, R0) || L(mem, R0));
    #4, Gamma_#4 := bvadd32(R0[32:0], 4294967295bv32), Gamma_R0;
    VF, Gamma_VF := bvnot1(bvcomp33(sign_extend1_32(bvadd32(#4, 1bv32)), bvadd33(sign_extend1_32(R0[32:0]), 0bv33))), (Gamma_R0 && Gamma_#4);
    CF, Gamma_CF := bvnot1(bvcomp33(zero_extend1_32(bvadd32(#4, 1bv32)), bvadd33(zero_extend1_32(R0[32:0]), 4294967296bv33))), (Gamma_R0 && Gamma_#4);
    ZF, Gamma_ZF := bvcomp32(bvadd32(#4, 1bv32), 0bv32), Gamma_#4;
    NF, Gamma_NF := bvadd32(#4, 1bv32)[32:31], Gamma_#4;
    assert Gamma_ZF;
    if ((bvnot1(bvcomp1(ZF, 1bv1)) != 0bv1)) {
      goto l00000937;
    }
    goto l0000094e;
  l00000937:
    R0, Gamma_R0 := zero_extend32_32(memory_load32_le(stack, bvadd64(R31, 12bv64))), gamma_load32(Gamma_stack, bvadd64(R31, 12bv64));
    goto l00000943;
  l0000094e:
    R0, Gamma_R0 := 0bv64, true;
    goto l00000943;
  l00000943:
    R31, Gamma_R31 := bvadd64(R31, 16bv64), Gamma_R31;
    return;
}
